docs/eventsV2.md

```diff
@@ -0,0 +1,268 @@
+# Queues V2 (updated May 8, 2022)
+
+## Design of Dark Queues V2
+
+Goals:
+
+- do not do DB queries to find events to run (that is, move scheduling outside DB)
+- support existing queue features (paused/blocked, delayed retries)
+- do not run a scheduling service
+
+## Definitions
+
+- `emit` adding an event to the queue from a Cron handler or the `emit` function
+- `event`: refers to the Dark value `emit`ed into the queue and the metadata around
+  it (including what you might call a "job" or "message" in other systems). We say we
+  "run the event" when we execute it, or "retry the event" if we requeue it.
+- `QueueWorker`: one of the machines which is running 1 or more events.
+- `queue`: encompasses everything else here
+
+## High level description
+
+The queue implements the following features:
+
+- events are added by calling `emit` or via CronChecker
+- events are run in `WORKER` handlers on the canvas with the same name that the event
+  was emitted to
+- events are run asynchronously on separate worker machines
+- worker handlers can be paused, meaning we won't start new events from that handler
+- worker handlers can be blocked by Dark admins to prevent operational issues
+- the UI shows the number of events in the queue
+- at least once execution (events which experience issues after partial execution are retried)
+- completed events are never retried (even if they complete with an error value,
+  including `incomplete` or a type error)
+- errors during execution cause events to be retried 5 minutes later (at most twice,
+  after which they are dropped)
+
+The queue also has the following accidental features:
+
+- events do not have an execution time-limit
+- events may be cancelled if the machine they are running on turns off. This should
+  only happen to events that run longer than 28 seconds.
```

Comment:
why 28?

Reply:
That's how long grace period a machine is given during shutdown.

Reply:
👍